src/Momento.Sdk/Internal/ControlGrpcManager.cs

```diff
@@ -1,32 +1,114 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
 using Grpc.Core;
 using Grpc.Core.Interceptors;
 using Grpc.Net.Client;
 using Microsoft.Extensions.Logging;
+using Momento.Protos.CacheClient;
 using Momento.Protos.ControlClient;
+using Momento.Sdk.Config.Middleware;
 using Momento.Sdk.Internal;
+using Momento.Sdk.Internal.Middleware;
 using static System.Reflection.Assembly;
 
 namespace Momento.Sdk.Internal;
 
+
+public interface IControlClient
+{
+    public Task<_CreateCacheResponse> CreateCacheAsync(_CreateCacheRequest request, CallOptions callOptions);
+    public Task<_DeleteCacheResponse> DeleteCacheAsync(_DeleteCacheRequest request, CallOptions callOptions);
+    public Task<_ListCachesResponse> ListCachesAsync(_ListCachesRequest request, CallOptions callOptions);
+}
+
+
+// Ideally we would implement our middleware based on gRPC Interceptors.  Unfortunately,
+// the their method signatures are not asynchronous. Thus, for any middleware that may
+// require asynchronous actions (such as our MaxConcurrentRequestsMiddleware), we would
+// end up blocking threads to wait for the completion of the async task, which would have
+// a big negative impact on performance. Instead, in this commit, we implement a thin
+// middleware layer of our own that uses asynchronous signatures throughout.  This has
+// the nice side effect of making the user-facing API for writing Middlewares a bit less
+// of a learning curve for anyone not super deep on gRPC internals.
+internal class ControlClientWithMiddleware : IControlClient
+{
+    private readonly IList<IMiddleware> _middlewares;
+    private readonly ScsControl.ScsControlClient _generatedClient;
+
+    public ControlClientWithMiddleware(ScsControl.ScsControlClient generatedClient, IList<IMiddleware> middlewares)
+    {
+        _generatedClient = generatedClient;
+        _middlewares = middlewares;
+    }
+
+    public async Task<_CreateCacheResponse> CreateCacheAsync(_CreateCacheRequest request, CallOptions callOptions)
+    {
+        var wrapped = await WrapWithMiddleware(request, callOptions, (r, o) => _generatedClient.CreateCacheAsync(r, o));
+        return await wrapped.ResponseAsync;
+    }
+
+    public async Task<_DeleteCacheResponse> DeleteCacheAsync(_DeleteCacheRequest request, CallOptions callOptions)
+    {
+        var wrapped = await WrapWithMiddleware(request, callOptions, (r, o) => _generatedClient.DeleteCacheAsync(r, o));
+        return await wrapped.ResponseAsync;
+    }
+
+    public async Task<_ListCachesResponse> ListCachesAsync(_ListCachesRequest request, CallOptions callOptions)
+    {
+        var wrapped = await WrapWithMiddleware(request, callOptions, (r, o) => _generatedClient.ListCachesAsync(r, o));
+        return await wrapped.ResponseAsync;
+    }
+
+    private async Task<MiddlewareResponseState<TResponse>> WrapWithMiddleware<TRequest, TResponse>(
```

Comment:
This function looks completely independent from this class other than `this._middlewares`. Is it copypasted from somewhere else? Would you like to instead move this function to be an extension on `IList<IMiddleware>`?

edit: yup, appears to be copypasted from DataGrpcManager. Can you DRY instead?

Reply:
sure.

Reply:
Addressed in https://github.com/momentohq/client-sdk-dotnet/pull/281/commits/18996672143fd259bc5b410b16e4ff9e743f5da9